### PR TITLE
Add PreExplodeEvents

### PR DESCRIPTION
**Pull request by SageSphinx63920**
This pr add PreXExplodeEvents to cancel an explosion before Minecraft calculates all the changes to prevent lag spikes

### CODE REVIEW

File: patches/api/0057-Add-PreExplodeEvents.patch [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
++    @NotNull
++    public static HandlerList getHandlerList() {
++        return handlers;
++    }
++}
+diff --git a/src/main/java/org/purpurmc/purpur/event/entity/GoatRamEntityEvent.java b/src/main/java/org/purpurmc/purpur/event/entity/GoatRamEntityEvent.java
+new file mode 100644
+index 0000000000000000000000000000000000000000..f62c14f3d4999e9112c1c73642aa337d97b94b5a
+--- /dev/null
++++ b/src/main/java/org/purpurmc/purpur/event/entity/GoatRamEntityEvent.java
```
Comment (JanTuck) [inline patches/api/0057-Add-PreExplodeEvents.patch]: What about this?

**Comment by JanTuck** [comment]
How is this different from this? https://hub.spigotmc.org/javadocs/bukkit/org/bukkit/event/entity/ExplosionPrimeEvent.html

**Comment by SageSphinx63920** [comment]
We talked about this on the Discord. This event gets called before any calculation of the radius is happing. If you spawn a creeper with a really big explosion radius and cancel the explosion event, the game still calculates every block that needs to be destroyed, which drops the TPS. You can cancel this event to prevent this

File: patches/api/0057-Add-PreExplodeEvents.patch (hunk 1 above)
Comment (SageSphinx63920) [reply patches/api/0057-Add-PreExplodeEvents.patch]: I don't know? I was just editing purpur like normal and after I rebuild my patches this appeared?

**Comment by JanTuck** [comment]
Alright thx.

**Comment by SageSphinx63920** [comment]
Do to Billy might not reviewing this, can somebody else do?

File: patches/server/0310-Add-PreExplodeEvents.patch [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+diff --git a/src/main/java/net/minecraft/world/level/Explosion.java b/src/main/java/net/minecraft/world/level/Explosion.java
+index 51d8271cb8fe246f4c11db8fbc061ed63f416da0..3707347ec139b3ac0010fe0460314fd593f58fb6 100644
+--- a/src/main/java/net/minecraft/world/level/Explosion.java
++++ b/src/main/java/net/minecraft/world/level/Explosion.java
+@@ -45,6 +45,7 @@ import org.bukkit.craftbukkit.event.CraftEventFactory;
+ import org.bukkit.event.entity.EntityExplodeEvent;
+ import org.bukkit.Location;
+ import org.bukkit.event.block.BlockExplodeEvent;
++import org.purpurmc.purpur.event.entity.PreEntityExplodeEvent;
```
Comment (granny) [inline patches/server/0310-Add-PreExplodeEvents.patch]: inline import

File: patches/api/0057-Add-PreExplodeEvents.patch [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
++
++/**
++ * Called before a block's explosion is processed
++ */
++public class PreBlockExplodeEvent extends BlockEvent implements Cancellable {
++    private static final HandlerList handlers = new HandlerList();
++    private boolean cancelled;
++    private final float yield;
++
++    public PreBlockExplodeEvent(@NotNull final Block what, final float yield) {
```
Comment (granny) [inline patches/api/0057-Add-PreExplodeEvents.patch]: might as well include the exploded blockstate as a parameter to the event like in `BlockExplodeEvent`. Include another constructor as well.

**Review by granny — CHANGES_REQUESTED** [review]
Since the params & everything is almost identical, have you thought about just extending the `BlockExplodeEvent` event instead of `BlockEvent`?

**Comment by granny** [comment]
@SageSphinx63920 any updates? do you need help rebasing the branch?

**Review by granny — APPROVED** [review]

**Pull request merged**